libbpf-tools/klockstat.c

```diff
@@ -0,0 +1,557 @@
+		if (errno || env->nr_stack_entries <= 0) {
+			warn("Invalid NR_STACKS: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	case 'S':
+		if (!parse_sorts(env, arg)) {
+			warn("Bad sort string: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	case 'd':
+		errno = 0;
+		duration = strtol(arg, NULL, 10);
+		if (errno || duration <= 0) {
+			warn("Invalid duration: %s\n", arg);
+			argp_usage(state);
+		}
+		env->duration = duration;
+		break;
+	case 'i':
+		errno = 0;
+		interval = strtol(arg, NULL, 10);
+		if (errno || interval <= 0) {
+			warn("Invalid interval: %s\n", arg);
+			argp_usage(state);
+		}
+		env->interval = interval;
+		break;
+	case 'R':
+		env->reset = true;
+		break;
+	case 'T':
+		env->timestamp = true;
+		break;
+	case 'h':
+		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+		break;
+	case ARGP_KEY_END:
+		if (env->duration) {
+			if (env->interval > env->duration)
+				env->interval = env->duration;
+			env->iterations = env->duration / env->interval;
+		}
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+struct stack_stat {
+	uint32_t stack_id;
+	struct lock_stat ls;
+	uint64_t bt[PERF_MAX_STACK_DEPTH];
+};
+
+static bool caller_is_traced(struct ksyms *ksyms, uint64_t caller_pc)
+{
+	const struct ksym *ksym;
+
+	if (!env.caller)
+		return true;
+	ksym = ksyms__map_addr(ksyms, caller_pc);
+	if (!ksym)
+		return true;
+	return strncmp(env.caller, ksym->name, strlen(env.caller)) == 0;
+}
+
+static int larger_first(uint64_t x, uint64_t y)
+{
+	if (x > y)
+		return -1;
+	if (x == y)
+		return 0;
+	return 1;
+}
+
+static int sort_by_acq(const void *x, const void *y)
+{
+	struct stack_stat *ss_x = *(struct stack_stat**)x;
+	struct stack_stat *ss_y = *(struct stack_stat**)y;
+
+	switch (env.sort_acq) {
+	case SORT_ACQ_MAX:
+		return larger_first(ss_x->ls.acq_max_time,
+				    ss_y->ls.acq_max_time);
+	case SORT_ACQ_COUNT:
+		return larger_first(ss_x->ls.acq_count,
+				    ss_y->ls.acq_count);
+	case SORT_ACQ_TOTAL:
+		return larger_first(ss_x->ls.acq_total_time,
+				    ss_y->ls.acq_total_time);
+	}
+
+	warn("bad sort_acq %d\n", env.sort_acq);
+	return -1;
+}
+
+static int sort_by_hld(const void *x, const void *y)
+{
+	struct stack_stat *ss_x = *(struct stack_stat**)x;
+	struct stack_stat *ss_y = *(struct stack_stat**)y;
+
+	switch (env.sort_hld) {
+	case SORT_HLD_MAX:
+		return larger_first(ss_x->ls.hld_max_time,
+				    ss_y->ls.hld_max_time);
+	case SORT_HLD_COUNT:
+		return larger_first(ss_x->ls.hld_count,
+				    ss_y->ls.hld_count);
+	case SORT_HLD_TOTAL:
+		return larger_first(ss_x->ls.hld_total_time,
+				    ss_y->ls.hld_total_time);
+	}
+
+	warn("bad sort_hld %d\n", env.sort_hld);
+	return -1;
+}
+
+static char *symname(struct ksyms *ksyms, uint64_t pc, char *buf, size_t n)
+{
+	const struct ksym *ksym = ksyms__map_addr(ksyms, pc);
+
+	if (!ksym)
+		return "Unknown";
+	snprintf(buf, n, "%s+0x%lx", ksym->name, pc - ksym->addr);
+	buf[n - 1] = 0;
```

Comment:
snprintf already handle this, no ?

Reply:
probably.  the man page doesn't say anything about that.  https://pubs.opengroup.org/onlinepubs/9699919799/functions/snprintf.html says it does.  my workstation's glibc does.  i've been burnt in the past with weird snprintfs, but i guess times have changed.  =)